### ISSUE TITLE
[Compatibility Version] Guidance on API version promotion

### ISSUE THREAD

**Issue by Jefftree** [kubernetes/kubernetes#127791]
Walking through a couple scenarios of how users would go about promoting their resources with the introduction of compatibility version. Feature gates are self explanatory via versioned_kube_features.go but resource APIs are a bit different. 

We determine the resources we serve based on the emulation version https://github.com/kubernetes/kubernetes/blob/master/pkg/controlplane/apiserver/apis.go#L91. Alpha APIs are disabled by default, and not compatible with emulation version so we can omit them from the current discussion. Beta APIs are also disabled by default, but is something we want to capture in emulation version. GA APIs are enabled by default, and is definitely something we want to capture in emulation version.

Promoting an API from alpha to beta and from beta to GA yields certain cases that need to be handled with emulation version.

User stories will be from the perspective of a feature developer. 

## User Story 1 : Promoting an API from alpha to beta.

Somewhere in the apiserver I have something like 

```go
// v1.31
if featuregate.Enabled(MyFeature) {
   r := &v1alpha1.MyType{}
   // operations on the resource
}
```

Once the feature goes beta, the traditional workflow will change the code to something like 

```go
// v1.32
if featuregate.Enabled(MyFeature) {
  // alpha resource directly replaced with beta resource on new version
   r := &v1beta1.MyType{}
   // operations on the resource
}
```

This is fine because we do not allow [alpha APIs to be enabled along with emulation version](https://github.com/kubernetes/enhancements/blob/master/keps/sig-architecture/4330-compatibility-versions/README.md#api-availability).  Thus, 1.32 can emulate 1.31 but `MyFeature` cannot be enabled, and despite not knowing about how to handle the alpha type when emulating 1.31, the code path will never be exercised. 

### User Story 2: Promoting feature from beta to GA

```go
// v1.32
if featuregate.Enabled(MyFeature) {
   r := &v1beta1.MyType{}
   // operations on the resource
}
```

```go
// v1.33
if featuregate.Enabled(MyFeature) {
   r := &v1.MyType{}
   // operations on the resource
}
```

Because of the [beta APIs off by default](https://github.com/kubernetes/enhancements/blob/master/keps/sig-architecture/3136-beta-apis-off-by-default/README.md), the feature gate will not be enabled by default in beta. However, let's say the API goes GA in 1.33. 1.33 cannot emulate 1.32 because the code operates on the GA types. v1 does not exist in 1.32's served versions so we cannot operate on it.

How would this be fixed? As a user would I need to perform something like this to satisfy compatibility version?

```go
// v1.33
if featuregate.Enabled(MyFeature) && featuregate.EmulationVersion() >= "1.33" {
  // alpha resource directly replaced with beta resource on new version
   r := &v1.MyType{}
   // operations on the resource
} else if  featuregate.Enabled(MyFeature) && featuregate.EmulationVersion() < "1.33" {
   r := &v1beta1.MyType{}
   // operations on the resource
}
```

This example is only for manipulation on a single type. If I'm writing a larger controller, would the entire controller code need to be duplicated and switched on emulation version? This seems like we're adding a bunch of extra work for beta to GA promotions (in addition to the complicated testing rules), and will need documentation if we pursue this path.

Alternatively, we can disallow emulation version to be set if beta features relying on beta APIs are enabled (they're off by default, so basically not allowing a default off feature to be set to default on under emulation version). Beta APIs are off by default due to stability risks and so we could mirror that with emulation version in not emulating unstable behavior. This will allow beta to GA transitions to be just like alpha to beta transitions in directly replacing the types. 

/sig api-machinery
/triage accepted

**Comment by jpbetz** [comment]
I don't see any way around having to support the beta version of API when emulating a controller that only had beta API support at the version being emulated.

But maybe we can find some pattern or offer some utilities to make it easier to support both API versions?

**Comment by VishalGawade1** [comment]
Is this issue still open? I would like to contribute
we can create a generic wrapper that can handle both beta and GA versions of a resource, simplifying the code needed in controllers and other components.
Let me know

**Comment by aaron-prindle** [comment]
How could we enforce this for users deprecating a Beta API?  Would this be part of prerelease-lifecycle-gen or potentially a new linter?  Is there single "source of truth" for each release that states each API and it's release lifecycle? (For feature gates, this is somewhat non-trivial for past releases with examining a releases kube_features.go - https://github.com/kubernetes/kubernetes/pull/127200#discussion_r1768862429)

**Comment by jpbetz** [comment]
> How could we enforce this for users deprecating a Beta API?

Is the new --emulate-version test your adding sufficient?  I was expecting that if the kube-apiserver and kube-controller-manager are both run with --emulate-version=1.31 (for binary version 1.32) then the v1beta1 branch of those conditionals would have to work for the test to pass (since the kube-apiserver won't serve v1).

**Comment by aaron-prindle** [comment]
Yes exactly, that is the goal of the test.  I didn't know how "shift-left" such an error might need to be but currently the plan for the `--emulate-version` test is that it is is a nightly that is initially not release blocking but eventually is release blocking once after X time of success (no flakes, etc.)

**Comment by Jefftree** [comment]
> I don't see any way around having to support the beta version of API when emulating a controller that only had beta API support at the version being emulated.
> 
> But maybe we can find some pattern or offer some utilities to make it easier to support both API versions?

Here's an example of one of my previous type promotion PRs. https://github.com/kubernetes/kubernetes/pull/122882 It touches the signature of many functions/loops/variable declarations. Will every single spot need to switch on a conditional for emulation version? I don't know the answer for this but depending on the feature it could be a lot of excessive work? If we say allow operations on the v1 resource but convert to v1beta1 when serving, the behavior is transparent to the user. However, there's no doubt that the code has changed and new bugs could have been introduced with the conversion. 

> Would this be part of prerelease-lifecycle-gen or potentially a new linter? Is there single "source of truth" for each release that states each API and it's release lifecycle? (For feature gates, this is somewhat non-trivial for past releases with examining a releases kube_features.go - https://github.com/kubernetes/kubernetes/pull/127200#discussion_r1768862429)

APIs are a bit different than features in that each one has its own group version and prerelease lifecycle tags. (instead of sharing a single variable toggled between alpha, beta, ga in the case of features). I *think* we can use those as the source of truth? We already use the emulation version in conjunction with those tags to determine which APIs to enable/disable.

> Is the new --emulate-version test your adding sufficient? I was expecting that if the kube-apiserver and kube-controller-manager are both run with --emulate-version=1.31 (for binary version 1.32) then the v1beta1 branch of those conditionals would have to work for the test to pass (since the kube-apiserver won't serve v1).

Since we haven't had a beta to GA type promotion in 1.32 yet I'd imagine the tests currently pass but may fail closer to code freeze. Maybe an interesting experiment we can do is to take off the guard rails and try to emulate 1.30? I'd imagine we'll see examples of those problems in prior releases.

> I wasn't sure if it might also be possible to bake something into prerelease-lifecycle-gen or what other options might be for enforcement earlier on in the process

What about yet another linter? :)

**Comment by jpbetz** [comment]
I'd be super curious how a linter for this might work.  It feels like it's logic in the code that is hard to identify with a lint rule, but maybe I'm wrong? Or maybe there is a convention we could establish that also happens to be easy to lint?

**Comment by Jefftree** [comment]
Siyuan had some interesting logic in the feature gate linter that diffs the unversioned feature list and errors out if something new is added (making it basically impossible for human error to add to the unversioned list unless they edit the yaml by hand). We could add something similar to have a yaml of the APIs and version based on the prerelease lifecycle and detect removals of APIs. The logic would then be that APIs may only be removed (deleted in code) if they've been marked as removed for 3 cycles already.

**Comment by sftim** [comment]
Relevant to https://github.com/kubernetes/website/issues/48171

**Comment by aojea** [comment]
> I don't see any way around having to support the beta version of API when emulating a controller that only had beta API support at the version being emulated.

I'm with Joe here, all the ways to support emulation seems to require adding code and wrap business logic that means it will be more difficult to debug or troubleshoot problems and much higher risk of having subtle bugs on those features that are already lacking exposure to the users since the "beta off by default", we may start conservative and not allow emulation on those

**Comment by siyuanfoundation** [comment]
@aojea Do you mean not allowing emulating beta apis that are disabled by default? Is it practical for upgrade? Beta apis could potentially have high adoption rates

**Comment by siyuanfoundation** [comment]
@VishalGawade1 can you share any examples of such wrappers? Thanks!

**Comment by aojea** [comment]
> Beta apis could potentially have high adoption rates

the problem here is not the beta api itself, is the combo of beta and feature gate that depends on that beta ... since those use to be complex features, we have DynamicResourceAllocation and ServiceCIDR that I'm aware of, and those require enable the feature gates, I don't think those will get high adoption rates ... and those that are adopting them already have to enable the feature gates to do that

**Comment by siyuanfoundation** [comment]
> > Beta apis could potentially have high adoption rates
> 
> the problem here is not the beta api itself, is the combo of beta and feature gate that depends on that beta ... since those use to be complex features, we have DynamicResourceAllocation and ServiceCIDR today that I'm aware of, and those require enable the feature gates, I don't think those will get high adoption rates ... and those that are adopting them already have to enable the feature gates to do that

But we do allow beta features to be emulated. So if the cluster starts to use the beta apis by enabling the beta feature gate, we need support them in emulated version.

**Comment by aojea** [comment]
> But we do allow beta features to be emulated. So if the cluster starts to use the beta apis by enabling the beta feature gate, we need support them in emulated version.

that is the paradox we have with the combo feature that depends on API, take a feature like ServiceCIDR that is exactly with @Jefftree is describing https://github.com/kubernetes/kubernetes/pull/128971

The feature and the API moves at the same time, the API without the feature (controllers, ...) is useless, why do you want to emulate an API like that? In addition, the cluster does not even start if the feature is not enabled, and if you enable it it fails becuse is looking for the GA API that does not exist in the previous version

Check the failure in the test `k8s.io/kubernetes/test/integration: etcd ` in https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/128971/pull-kubernetes-integration/1873855559963250688 , the test `=== RUN   TestEtcdStoragePath/1.31` fails  to start because is not able to find the GA APIS , needed by the controllers to work

**Comment by siyuanfoundation** [comment]
> 
> to be more clear, the problem is not to emulate the APIs is to emulate the feature gates that depend on the APIs
> 
> Feature Gate	API	Binary Version	Emulated Version	Result
> GA	v1	1.33	-	ok
> Beta	v1beta1	1.33	1.32	controller depends on v1 and fails

Isn't the problem in this case that the controller should handle both versions of API, like @Jefftree described? 
```
// v1.33
if featuregate.Enabled(MyFeature) && featuregate.EmulationVersion() >= "1.33" {
  // alpha resource directly replaced with beta resource on new version
   r := &v1.MyType{}
   // operations on the resource
} else if  featuregate.Enabled(MyFeature) && featuregate.EmulationVersion() < "1.33" {
   r := &v1beta1.MyType{}
   // operations on the resource
}
```

If we do not allow emulating complex features like ServiceCIDR, it would violate our goal in the KEP to support emulating Beta features. And if the user starts to use one of these features, they would not be able to upgrade through the safe upgrade path (binary 1.32 -> [binary 1.33, emulated 1.32] -> [binary 1.33, emulated 1.33] , which means we would need to have a separate upgrade policy depending on the features/APIs a cluster has turned on. 
Even if we can do that, we still need to handle the graduations of APIs not gated by a feature gate, which would still require adding code and wrap business logic.

**Comment by siyuanfoundation** [comment]
> 
> AFAIK APIs are not feature gated and they should be using a stored version n-1 , so there should not be any need for handling the APIs
> 
Does the storage version policy of n-1 need be changed to n-3 for emulation version then?

**Comment by Jefftree** [comment]
> The KEP is to make upgrades safer based on the hypothesis that being able to emulate features we are able to make this safe path, but we are emulating business logic and duplicating code paths that can drift or have copy paste errors or will need backports ... debugging and troubleshooting will be really complex and will add a lot of overhead to the feature owners ... I'm skeptical that this approach will make the upgrade safer than with today's methid since we are "faking" the emulation and we may have copy paste errors ... I don't think there is enough return on investment

I agree to a certain extent, if we are duplicating a lot of business logic and code paths, the overhead will make it difficult for maintenance. That's the exact reason we're not emulating alphas because they're more unstable and business logic could change drastically between alphas and betas describing the exact scenario you mentioned with the overhead. For beta->GA, generally the changes in business logic and API is more contained and in a lot of scenarios almost a duplicate of the types from beta -> GA. Thus, I think it's worth an effort for us to try to support beta->GA rather than dropping support for complex features that depend on API. A quick scan through https://github.com/kubernetes/kubernetes/pull/128971 seems to suggest mostly a `s/v1beta1/v1` operation for beta->GA type which I feel like we *should* be able to safely emulate with the right changes.

> Does the storage version policy of n-1 need be changed to n-3 for emulation version then?

With compatibility versions, storage version defaults to `minCompatVersion` which is `emulatedVersion - 1` == `n-1` by default. Given the two parameters, we're technically widening the net to support `{n-3,...,n}` (n because we could set `minCompatVersion = emulationVersion` and disallow downgrades). I think we could update the documentation to indicate this range of support?

**Comment by aojea** [comment]
> For beta->GA, generally the changes in business logic and API is more contained and in a lot of scenarios almost a duplicate of the types from beta -> GA. Thus, I think it's worth an effort for us to try to support beta->GA rather than dropping support for complex features that depend on API. A quick scan through #128971 seems to suggest mostly a `s/v1beta1/v1` operation for beta->GA type which I feel like we _should_ be able to safely emulate with the right changes.

ServiceCIDR is a very simple API ... if you check at DRA you'll see that is not as straightforward ... if we can not generalize that, we can not be consistent and then emulating a beta feature that is disabled by default does not provide enough benefit vs the cost of maintenance and the developer experience ...

**Comment by siyuanfoundation** [comment]
Since 
1. it is impractical to use switch statements everywhere in the code.
2. it seems very difficult to design a generic enough interface/wrapper around the versions
We will probably need to emulate serving the new API somehow. 
If that is the case, why don't we just serve the new APIs in the apiserver?

We would need to reframe the purpose of emulated version for APIs: it is more about keeping the removed APIs (Beta and GA) in emulation mode. As for API versions introduced after the emulated version, it is ok to serve and use them because the underlying storage version is the same. This should be ok since most likely Beta->GA APIs do not change much. 

WDYT? @jpbetz @liggitt @aojea
